pandas/core/arrays/masked.py

```diff
@@ -0,0 +1,117 @@
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from pandas.core.dtypes.common import is_integer
+from pandas.core.dtypes.missing import isna, notna
+
+from pandas.core.algorithms import take
+from pandas.core.arrays import ExtensionArray, ExtensionOpsMixin
+import pandas.core.common as com
+from pandas.core.indexers import check_bool_array_indexer
+
+if TYPE_CHECKING:
+    from pandas._typing import Scalar
+
+
+class BaseMaskedArray(ExtensionArray, ExtensionOpsMixin):
+    """
+    Base class for masked arrays (which use _data and _mask to store the data).
+
+    numpy based
+    """
+
+    _data: np.ndarray
+    _mask: np.ndarray
+
+    # The value used to fill '_data' to avoid upcasting
+    _internal_fill_value: "Scalar"
+
+    def __getitem__(self, item):
+        if is_integer(item):
+            if self._mask[item]:
+                return self.dtype.na_value
+            return self._data[item]
+
+        elif com.is_bool_indexer(item):
```

Comment:
Possibly out of scope, but I see the same optimization issue we had with Categorical. Can you convert `item` to an ndarray before doing this `elif`? Or do you want to keep this as a straight refactor, and do that as a followup?

Reply:
Yes, I also noted that on the PR (https://github.com/pandas-dev/pandas/pull/30747#issuecomment-571278131), I think it is still worth thinking this through if we need a more general "check" function (https://github.com/pandas-dev/pandas/issues/30744#issuecomment-571276333, #30738)